app/code/community/Payone/Core/Model/Service/TransactionStatus/Execute.php

```diff
@@ -32,160 +32,234 @@
+        }
+
+        if (!empty($this->excludedTxIds)) {
+            $collection->addFieldToFilter('txid', array('nin' => $this->excludedTxIds));
+        }
+
+        // Make sure we get the most recent TX status to make sure newer status are processed
+        // before any old dangling status.
+        $collection->setOrder('txtime', Payone_Core_Model_Domain_Resource_Protocol_TransactionStatus_Collection::SORT_ORDER_DESC);
+
+        // Limit result to one entity
+        $collection->getSelect()->limit(1);
+
+        $this->logMessage(sprintf('Fetching next TX status -- %s', $collection->getSelectSql(true)));
+
+        $collection->load();
+
+        /** @var Payone_Core_Model_Domain_Protocol_TransactionStatus $txStatus */
+        foreach ($collection as $txStatus) {
+            return $txStatus;
+        }
+
+        return null;
     }
 
-    protected function _getIncrementId($sReference)
+    /**
+     * Verifies the provided TX status.
+     *
+     * @param Payone_Core_Model_Domain_Protocol_TransactionStatus $txStatus
+     * @return bool
+     */
+    protected function verifyTxStatus(Payone_Core_Model_Domain_Protocol_TransactionStatus $txStatus)
     {
-        $oResource = Mage::getSingleton('core/resource');
-        $oRead = $oResource->getConnection('core_read');
+        $collection = $this->getFactory()->getModelTransactionStatus()->getCollection();
 
-        $select = $oRead->select()
-            ->from(array('tbl' => $oResource->getTableName('sales/order')), array('increment_id'))
-            ->where('payone_cancel_substitute_increment_id = ?', $sReference)
-            ->order('entity_id desc')
-            ->limit(1);
-        $sSubstitudeIncrementId = $oRead->fetchOne($select);
-        if (!empty($sSubstitudeIncrementId)) {
-            return $sSubstitudeIncrementId;
+        // Get all prior TX status with same TxId
+        $collection->clear();
+        $collection->addFieldToFilter('id', array('neq' => (int) $txStatus->getId()));
+        $collection->addFieldToFilter('txid', (int) $txStatus->getTxid());
+        $collection->addFieldToFilter('txtime', array('lteq' => (int) $txStatus->getTxtime()));
+        $collection->setOrder('txtime', Payone_Core_Model_Domain_Resource_Protocol_TransactionStatus_Collection::SORT_ORDER_ASC);
+
+        $this->logMessage(sprintf("Verify TX status %d (%d) prior TX status are all processed successfully.", $txStatus->getId(), $txStatus->getTxid()));
+        $this->logMessage(sprintf("Fetch prior TX status -- %s", $collection->getSelectSql(true)));
+
+        $collection->load();
+
+        // Verify all prior TX status have a processing_status of completed
+        /** @var Payone_Core_Model_Domain_Protocol_TransactionStatus $priorTxStatus */
+        foreach ($collection as $priorTxStatus) {
+            if ($priorTxStatus->getProcessingStatus() !== Payone_Core_Model_Domain_Protocol_TransactionStatus::STATUS_OK) {
```

Comment:
Can't we also filter the collection by this and then just check the size of the resulting collection?

Reply:
Hi @sprankhub ,

what do you mean with 'this' @
> filter the collection by this

Greeting from Kiel
Timo

Reply:
Well, you get the full collection of all tx statuses and then, when iterating over them, you filter out some of them. It would be more efficient to set this filter (by processing status) directly to the collection. But maybe I miss something here - I just had a short look...

Reply:
Hey @sprankhub ,

you are generally right here. The verifyTxStatus() is basically for ensuring the provided TX status is ready for processing. That said, the method may contain any arbitrary logic (in the future) to allow / prohibit processing of the provided TX status. For sake of clarity I would like to leave the logic encapsulated in this method, just like it is right now.

Best regards,
Fabian